src/StatusQ/Controls/StatusPinInput.qml

```diff
@@ -0,0 +1,137 @@
+
+    property string pinInput: inputText.text
+    property var validator: inputText.validator
+    property int pinLen: 6
+    property int circleSpacing: 16
+    property int circleDiameter: 16
+
+    QtObject {
+        id: d
+        property int currentPinIndex: 0
+    }
+
+    function statesInitialization() {
+        d.currentPinIndex = 0
+        repeater.itemAt(d.currentPinIndex).innerState = "NEXT"
+        for (var i = 1; i < root.pinLen; i++) {
+            repeater.itemAt(i).innerState = "EMPTY"
+        }
+    }
+
+    width: (root.circleDiameter + root.circleSpacing) * root.pinLen
+    height: root.circleDiameter
+
+    Component.onCompleted: {
+        statesInitialization()
+        inputText.forceActiveFocus()
+    }
+
+    // Pin input data management object:
+    TextInput {
+        id: inputText
+        visible: false
+        focus: true
+        maximumLength: root.pinLen
+        validator: root.validator
+        onTextChanged: {
+            // Modify state of current introduced character position:
+            if(text.length >= (d.currentPinIndex + 1)) {
+                repeater.itemAt(d.currentPinIndex).innerState = "FILLED"
+
+                // Update next:
+                d.currentPinIndex++
+                if(d.currentPinIndex < root.pinLen)
+                    repeater.itemAt(d.currentPinIndex).innerState = "NEXT"
+
+            }
+            // Modify state of current removed character position:
+            else if (text.length <= (d.currentPinIndex + 1)) {
+                if(d.currentPinIndex < root.pinLen)
+                    repeater.itemAt(d.currentPinIndex).innerState = "EMPTY"
+                d.currentPinIndex--
+                repeater.itemAt(d.currentPinIndex).innerState = "NEXT"
+            }
+
+            // Some component validations:
+            if(text.length !== d.currentPinIndex)
+                console.error("StatusPinInput input management error. Current pin length must be "+ text.length + "and is " + d.currentPinIndex)
+        }
+    }
+
+    // Pin input visual objects:
+    Row {
+        spacing: root.circleSpacing
+
+        Repeater {
+            id: repeater
+            model: root.pinLen
+
+            Rectangle {
+                id: background
+                property string innerState: "EMPTY"
+
+                 width: root.circleDiameter
+                 height: width
+                 color: Theme.palette.primaryColor2
+                 radius: 0.5 * width
+
+                 Rectangle {
+                     id: inner
+                     state: background.innerState
+                     anchors.centerIn: parent
+                     height: width
+                     color: Theme.palette.primaryColor1
+                     radius: 0.5 * width
+
+                     states: [
+                         State {
+                             name: "NEXT"
+                             PropertyChanges {target: inner; width: root.circleDiameter / 2}
```

Comment:
since all states are changing the same property and using the same duration (for transitions) would it make sense to use inline javascript instead and have `Behavior on width { NumberAnimation { duration: 200 } }` do the job?

Reply:
Totally right! Changed!